api/protocol.go

```diff
@@ -52,68 +50,148 @@ type Response struct {
 	Data    map[string]interface{} `json:"data,omitempty"`
 }
 
-// ReadMessage reads a message from reader. A message is either a Request or a
-// Response
-func ReadMessage(reader io.Reader, msg interface{}) error {
-	buf := make([]byte, headerLength)
-	n, err := reader.Read(buf)
+func maxOpcodeForFrameType(t FrameType) int {
+	switch t {
+	default:
+		fallthrough
+	case TypeCommand:
+		return int(CmdMax)
+	case TypeResponse:
+		return int(CmdMax)
+	case TypeStream:
+		return int(StreamMax)
+	case TypeNotification:
+		return int(NotificationMax)
+	}
+}
+
+// ReadFrame reads a full frame (header and payload) from r.
+func ReadFrame(r io.Reader) (*Frame, error) {
+	// Read the header.
+	buf := make([]byte, minHeaderLength)
+	n, err := r.Read(buf)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if n != headerLength {
-		return errors.New("couldn't read the full header")
+	if n != minHeaderLength {
+		return nil, errors.New("frame: couldn't read the full header")
 	}
 
-	header := header{
-		length: binary.BigEndian.Uint32(buf[0:4]),
-		flags:  binary.BigEndian.Uint32(buf[4:8]),
+	// Decode it.
+	frame := &Frame{}
+	header := &frame.Header
+	header.Version = int(binary.BigEndian.Uint16(buf[0:2]))
+	if header.Version < 2 || header.Version > Version {
+		return nil, fmt.Errorf("frame: bad version %d", header.Version)
+	}
+	header.HeaderLength = int(buf[2]) * 4
+	header.Type = FrameType(buf[6] & 0xf)
+	flags := buf[6] & 0xf0
```

Comment:
@dlespiau Shouldn't this be buf[6] & 0x10 for a stricter check since the ascii diagram in the docs shows that the 5th bit is set for error and all other bits are reserved, I have been OR'ing with 0x10, and not seeing the right results when error is supposed to be set. Seems that the 6th bit is being set in case of error (eg. 0010 0001 and not 
0001 0001) as expected. Either the docs need to be changed to mention the actual bit/bits to check/unreserve or this needs to be corrected.